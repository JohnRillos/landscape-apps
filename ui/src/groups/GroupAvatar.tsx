import cn from 'classnames';
import React from 'react';
import ColorBoxIcon from '@/components/icons/ColorBoxIcon';
import { isColor } from '@/logic/utils';
import useMedia from '@/logic/useMedia';
import { useCalm } from '@/state/settings';

interface GroupAvatarProps {
  image?: string;
  size?: string;
  className?: string;
  title?: string;
}

const textSize = (size: string) => {
  const dims = parseInt(size.replace(/[^0-9.]/g, ''), 10);
  switch (dims) {
    case 7272:
      return 'text-3xl';
    case 2020:
      return 'text-xl';
    case 1616:
      return 'text-xl';
    case 1414:
      return 'text-xl';
    case 1212:
      return 'text-xl';
    case 66:
      return 'text-sm';
    default:
      return 'text-sm';
  }
};

export default function GroupAvatar({
  image,
  size = 'h-6 w-6',
  className,
  title,
}: GroupAvatarProps) {
  const dark = useMedia('(prefers-color-scheme: dark)');
  const calm = useCalm();
  let background;

  if (!calm.disableRemoteContent) {
    background = image || (dark ? '#333333' : '#E5E5E5');
  } else {
    background = dark ? '#333333' : '#E5E5E5';
  }

  return image && !calm.disableRemoteContent && !isColor(image) ? (
    <img className={cn('rounded', size, className)} src={image} />
  ) : (
    <ColorBoxIcon
      className={cn('rounded', size, textSize(size), className)}
      color={background}
      letter={title ? title.charAt(0) : ''}
    />
  );
}
