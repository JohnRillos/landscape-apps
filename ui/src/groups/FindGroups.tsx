import React, { useCallback, useEffect, useState } from 'react';
import { useGroupState, usePendingGangs } from '@/state/groups';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import { Gangs } from '@/types/groups';
import useRequestState from '@/logic/useRequestState';
import GroupJoinList from './GroupJoinList';

// TODO: how to handle the invite link case?

export default function FindGroups() {
  const [foundGangs, setFoundGangs] = useState<Gangs | null>(null);
  const pendingGangs = usePendingGangs();
  const [shipSelectorShips, setShipSelectorShips] = useState<ShipOption[]>([]);
  const selectedShip =
    shipSelectorShips.length > 0 ? shipSelectorShips[0] : null;
  const presentedShip = selectedShip
    ? selectedShip.label || selectedShip.value
    : '';
  const { isPending, isReady, setPending, setReady } = useRequestState();

  const searchGroups = useCallback(async () => {
    if (!selectedShip) {
      return;
    }
    setFoundGangs(null);
    setPending();
    const results = await useGroupState.getState().index(selectedShip.value);
    setFoundGangs(results);
    setReady();
  }, [selectedShip, setPending, setReady]);

  useEffect(() => {
    if (shipSelectorShips.length > 0) {
      searchGroups();
    }
  }, [searchGroups, shipSelectorShips]);

  return (
    <div className="flex grow bg-gray-50">
      <div className="w-full max-w-3xl p-4">
        <section className="card mb-4 space-y-8 p-8">
          <h1 className="text-lg font-bold">Find Groups</h1>
          <div>
            <label htmlFor="flag" className="mb-1.5 block font-semibold">
              Join Groups via Nickname or Urbit ID
            </label>
            <div className="flex flex-col space-y-2">
              <ShipSelector
                ships={shipSelectorShips}
                setShips={setShipSelectorShips}
                isMulti={false}
                isClearable={true}
                isLoading={isPending}
                hasPrompt={false}
                placeholder={''}
              />
            </div>
          </div>
          {selectedShip ? (
            <section className="card mb-4 space-y-8 p-8">
              <p className="font-semibold text-gray-400">
                {isPending ? (
                  <>
                    <span>Searching for groups hosted by&nbsp;</span>
                    <span className="text-gray-800">{presentedShip}</span>
                    <span>...</span>
                  </>
                ) : isReady ? (
                  <>
                    <span>Groups hosted by&nbsp;</span>
                    <span className="text-gray-800">{presentedShip}</span>
                    <span>:</span>
                  </>
                ) : null}
              </p>
              {isReady && foundGangs ? (
                <GroupJoinList gangs={foundGangs} />
              ) : isReady ? (
                <p>No groups found for &apos;{presentedShip}&apos;</p>
              ) : null}
            </section>
          ) : null}
        </section>
        {Object.keys(pendingGangs).length > 0 ? (
          <section className="card mb-4 space-y-8 p-8">
            <h1 className="text-lg font-bold">Pending Invites</h1>
            <GroupJoinList gangs={pendingGangs} />
          </section>
        ) : null}
      </div>
    </div>
  );
}
