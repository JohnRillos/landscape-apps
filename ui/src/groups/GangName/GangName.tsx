import React from 'react';
import cn from 'classnames';
import { useGang } from '../../state/groups/groups';

export default function GangName(props: { flag: string; className?: string }) {
  const { flag, className } = props;
  const { preview } = useGang(flag);

  return (
    <span className={cn(className, preview ? '' : 'text-mono')}>
      {preview ? preview.meta.title : flag}
    </span>
  );
}
