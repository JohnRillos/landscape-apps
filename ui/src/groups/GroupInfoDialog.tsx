import React from 'react';
import GroupAvatar from './GroupAvatar';
import { useGroup, useRouteGroup } from '../state/groups';
import { useDismissNavigate } from '../logic/routing';
import Dialog, {
  DialogClose,
  DialogContent,
  DialogTitle,
} from '../components/Dialog';

export default function GroupInfoDialog() {
  const dismiss = useDismissNavigate();
  const flag = useRouteGroup();
  const group = useGroup(flag);
  const meta = group?.meta;

  const onOpenChange = (isOpen: boolean) => {
    if (!isOpen) {
      dismiss();
    }
  };

  if (!meta) {
    return null;
  }

  return (
    <Dialog defaultOpen onOpenChange={onOpenChange}>
      <DialogContent className="w-full sm:max-w-lg">
        <header className="flex items-center">
          <DialogTitle className="text-lg font-bold">Group Info</DialogTitle>
        </header>
        <div className="mt-6">
          <div className="flex flex-col items-center">
            <GroupAvatar img={meta.image} size="h-20 w-20" />
            <div className="my-4 text-center">
              <h2 className="center mb-2 font-semibold">{meta.title}</h2>
              {/* Current group meta object doesn't contain public/private info  */}
              <h3 className="text-base text-gray-600">Private Group</h3>
            </div>
            <p className="w-full leading-5">{meta.description}</p>
          </div>
        </div>
        <footer className="mt-8 flex items-center">
          <DialogClose asChild>
            <button className="button ml-auto">Close</button>
          </DialogClose>
        </footer>
      </DialogContent>
    </Dialog>
  );
}
