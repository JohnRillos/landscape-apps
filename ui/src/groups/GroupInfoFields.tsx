import React, { useEffect, useState } from 'react';
import { useFormContext } from 'react-hook-form';
import ColorPicker from '@/components/ColorPicker';
import ColorBoxIcon from '@/components/icons/ColorBoxIcon';
import EmptyIconBox from '@/components/icons/EmptyIconBox';
import useMedia from '@/logic/useMedia';
import { isValidUrl } from '@/logic/utils';
import { GroupMeta } from '@/types/groups';
import GroupAvatar from './GroupAvatar';

export default function GroupInfoFields() {
  const {
    register,
    unregister,
    watch,
    setValue,
    formState: { errors },
  } = useFormContext<GroupMeta>();
  const [iconType, setIconType] = useState<'image' | 'color'>();
  const [iconColor, setIconColor] = useState<string>();
  const [iconLetter, setIconLetter] = useState<string>();
  const [iconUrl, setIconUrl] = useState<string>();
  const watchIconColor = watch('color');
  const watchIconImage = watch('image');
  const watchTitle = watch('title');
  const dark = useMedia('(prefers-color-scheme: dark)');
  const defaultColor = dark ? '#FFFFFF' : '#000000';

  useEffect(() => {
    if (watchIconImage) {
      setIconType('image');
    } else if (watchIconColor) {
      setIconType('color');
    } else {
      setIconType(undefined);
    }
  }, [watchIconColor, watchIconImage]);

  useEffect(() => {
    if (iconType === 'color' && watchIconColor !== '') {
      setIconColor(watchIconColor as string);
    }
  }, [iconType, watchIconColor]);

  useEffect(() => {
    if (iconType === 'color' && watchTitle !== '') {
      setIconLetter((watchTitle as string).slice(0, 1));
    }
  }, [iconType, watchTitle]);

  useEffect(() => {
    if (iconType === 'image' && watchIconImage !== '') {
      setIconUrl(watchIconImage as string);
    }
  }, [iconType, watchIconImage]);

  useEffect(() => {
    register('color');
  }, []); // eslint-disable-line

  const handleCancelColorIcon = (e: React.SyntheticEvent) => {
    e.preventDefault();
    setIconType(undefined);
    setIconColor(undefined);
    setIconLetter(undefined);
    setValue('color', '');
  };

  const handleCancelImageIcon = (e: React.SyntheticEvent) => {
    e.preventDefault();
    setIconType(undefined);
    setIconUrl(undefined);
    setValue('image', '');
    unregister('image');
  };

  console.log(watchIconImage, iconUrl, isValidUrl(iconUrl));

  return (
    <>
      <div className="flex justify-between">
        <div className="flex flex-col">
          <span className="pb-2 font-bold">Group Icon *</span>
          <div className="flex items-center space-x-2">
            {iconType === undefined ? (
              <>
                <button
                  className="secondary-button"
                  onClick={() => setIconType('image')}
                >
                  Image URL
                </button>
                <span className="text-sm font-semibold">or</span>
                <button
                  className="secondary-button"
                  onClick={() => setIconType('color')}
                >
                  Fill Color
                </button>
              </>
            ) : null}
            {iconType === 'image' ? (
              <>
                <input
                  className="input"
                  placeholder="Paste Image URL"
                  {...register('image', {
                    required: true,
                    validate: (value) => isValidUrl(value),
                  })}
                  type="url"
                />
                <button
                  className="secondary-button"
                  onClick={handleCancelImageIcon}
                >
                  Cancel
                </button>
              </>
            ) : null}
            {iconType === 'color' ? (
              <div className="flex items-center space-x-2">
                <ColorPicker
                  color={
                    (watchIconColor as string) === ''
                      ? defaultColor
                      : (watchIconColor as string)
                  }
                  setColor={(newColor: string) =>
                    setValue('color', newColor || '', {
                      shouldDirty: true,
                      shouldTouch: true,
                    })
                  }
                />
                <button
                  className="secondary-button"
                  onClick={handleCancelColorIcon}
                >
                  Cancel
                </button>
              </div>
            ) : null}
            {errors.color ? (
              <span className="text-sm">{errors.color.message}</span>
            ) : null}
          </div>
        </div>
        {iconType === 'color' ? (
          <ColorBoxIcon
            className="h-12 w-12 text-xl"
            color={iconColor ? iconColor : '#000000'}
            letter={iconLetter ? iconLetter : 'T'}
          />
        ) : null}
        {iconType === 'image' && isValidUrl(iconUrl) ? (
          <GroupAvatar size="h-14 w-14" image={iconUrl} />
        ) : null}
        {iconType === undefined ? (
          <EmptyIconBox className="h-14 w-14 text-gray-300" />
        ) : null}
      </div>
      <div className="flex flex-col">
        <label htmlFor="title" className="pb-2 font-bold">
          Group Name *
        </label>
        <input
          // TODO: set sane maxLength
          {...register('title', { required: true, maxLength: 180 })}
          className="input"
          type="text"
          placeholder="Title"
        />
      </div>
      <div className="flex flex-col">
        <label htmlFor="description" className="pb-2 font-bold">
          Group Description (optional)
        </label>
        <textarea
          // TODO: set sane maxLength
          {...register('description', { maxLength: 300 })}
          className="input"
        />
      </div>
    </>
  );
}
