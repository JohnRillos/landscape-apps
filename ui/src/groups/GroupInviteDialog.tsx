import React, { useCallback } from 'react';
import Dialog, { DialogClose, DialogContent } from '../components/Dialog';
import { useDismissNavigate } from '../logic/routing';
import { useRouteGroup } from '../state/groups';

export default function GroupInviteDialog() {
  const dismiss = useDismissNavigate();
  const flag = useRouteGroup();

  const onOpenChange = (isOpen: boolean) => {
    if (!isOpen) {
      dismiss();
    }
  };

  const onInvite = useCallback(
    (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => null,
    []
  ); // TODO: invite poke

  return (
    <Dialog defaultOpen onOpenChange={onOpenChange}>
      <DialogContent containerClass="max-w-md" showClose>
        <div className="flex flex-col">
          <h2 className="mb-4 text-lg font-bold">Invite To Group</h2>
          <p className="mb-7 leading-5">Enter a username to invite to {flag}</p>
          <div className="flex items-center justify-end space-x-2">
            <DialogClose className="button" type="button">
              Cancel
            </DialogClose>

            <DialogClose
              onClick={onInvite}
              className="button bg-blue text-white"
            >
              Invite
            </DialogClose>
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
}
