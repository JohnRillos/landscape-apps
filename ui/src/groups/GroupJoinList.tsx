import React, { useCallback } from 'react';
import { useNavigate } from 'react-router';
import { useGang, useGroup, useGroupState } from '@/state/groups';
import GroupSummary, { GroupSummarySize } from './GroupSummary';

interface GroupJoinItemProps {
  flag: string;
  size?: GroupSummarySize;
}

function GroupJoinItem({ flag, size }: GroupJoinItemProps) {
  const navigate = useNavigate();
  const gang = useGang(flag);
  const group = useGroup(flag);

  const join = useCallback(() => {
    if (group) {
      return navigate(`/groups/${flag}`);
    }

    return useGroupState.getState().join(flag, true);
  }, [flag, group, navigate]);

  return (
    <li className="flex items-center p-2">
      <GroupSummary flag={flag} {...gang.preview} size={size} />
      <button
        className="button ml-auto bg-blue-soft text-blue dark:bg-blue-900"
        onClick={join}
      >
        {group ? 'Open' : 'Join'}
      </button>
    </li>
  );
}
interface GroupJoinListProps {
  gangs: string[];
  size?: GroupSummarySize;
}

export default function GroupJoinList({ gangs, size }: GroupJoinListProps) {
  return (
    <ul>
      {gangs.map((g) => (
        <GroupJoinItem key={g} flag={g} size={size} />
      ))}
    </ul>
  );
}
