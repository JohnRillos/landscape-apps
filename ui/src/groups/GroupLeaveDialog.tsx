import React, { useCallback } from 'react';
import Dialog, { DialogClose, DialogContent } from '@/components/Dialog';
import { useDismissNavigate } from '@/logic/routing';
import { useGroupState, useRouteGroup } from '@/state/groups/groups';

const { ship } = window;

export default function GroupInviteDialog() {
  const dismiss = useDismissNavigate();
  const flag = useRouteGroup();

  const onOpenChange = (isOpen: boolean) => {
    if (!isOpen) {
      dismiss();
    }
  };

  const onLeave = useCallback(() => {
    useGroupState.getState().leave(flag);
  }, [flag]);

  return (
    <Dialog defaultOpen onOpenChange={onOpenChange}>
      <DialogContent containerClass="w-full max-w-lg" showClose>
        <div className="flex flex-col">
          <h2 className="mb-4 text-lg font-bold">Leave Group</h2>
          <div className="w-full py-3">
            {flag.includes(ship) ? (
              <span className="text-sm font-semibold">
                You cannot leave a group that you host. You should delete the
                group instead
              </span>
            ) : (
              <span className="text-sm font-semibold">
                Do you really want to leave {flag}?
              </span>
            )}
          </div>
          <div className="flex items-center justify-end space-x-2">
            <DialogClose className="secondary-button">Cancel</DialogClose>
            {flag.includes(ship) ? null : (
              <DialogClose
                onClick={onLeave}
                className="button bg-blue text-white dark:text-black"
              >
                Leave
              </DialogClose>
            )}
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
}
