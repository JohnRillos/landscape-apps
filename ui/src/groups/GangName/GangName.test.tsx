import { describe, it, expect } from 'vitest';
import React from 'react';
import { render } from '@testing-library/react';
import GangName from './GangName';
import { mockGangs } from '../../mocks/groups';
import { useGroupState } from '../../state/groups/groups';

describe('ChatInput', () => {
  it('renders default', () => {
    const { asFragment } = render(<GangName flag="~zod/structure" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders preview', () => {
    useGroupState.getState().set((draft) => {
      draft.gangs = mockGangs;
    });
    const { asFragment } = render(<GangName flag="~zod/structure" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
