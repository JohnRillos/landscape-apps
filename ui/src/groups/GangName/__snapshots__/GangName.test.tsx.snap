// Vitest Snapshot v1

exports[`ChatInput > renders default 1`] = `
<DocumentFragment>
  <span
    class="text-mono"
  >
    ~zod/structure
  </span>
</DocumentFragment>
`;

exports[`ChatInput > renders preview 1`] = `
<DocumentFragment>
  <span
    class=""
  >
    Structure
  </span>
</DocumentFragment>
`;
